tempesta_fw/apm.c

```diff
@@ -0,0 +1,734 @@
+		tfw_stats_adjust(rng, 0);
+		return;
+	}
+
+	pc3.atomic = rng->ctl[3].atomic;
+	if (unlikely(r_time > pc3.end)) {
+		tfw_stats_extend(rng, r_time);
+		pc3.atomic = rng->ctl[3].atomic;
+	}
+	atomic_inc(__rng(&pc3, rng->cnt[3], r_time));
+	tfw_stats_adjust(rng, 3);
+}
+
+
+/*
+ * APM ring buffer.
+ *
+ * It consists of the predefined number of entries that are "reused" as
+ * the buffer gets full. The ring buffer as a whole keeps the APM stats
+ * for the latest time interval (the time window), and each entry of
+ * the buffer keeps the APM stats for a piece of that time (an interval).
+ */
+/*
+ * A ring buffer entry structure.
+ * @pcntrng	Struct for response time data by the percentiles algorithm.
+ * @jtstamp	Time the ring buffer entry has been started, in jiffies.
+ * @reset	The entry can be reset by one thread at a time.
+ */
+typedef struct {
+	TfwPcntRanges	pcntrng;
+	unsigned long	jtstamp;
+	atomic_t	reset;
+} TfwApmRBufEntry;
+
+/*
+ * The ring buffer structure.
+ * @rbent	Array of ring buffer entries.
+ * @rbentsz	The size of @rbent.
+ */
+typedef struct {
+	TfwApmRBufEntry *rbent;
+	int		rbentsz;
+} TfwApmRBuf;
+
+/*
+ * APM Data structure.
+ * @rbuf	The ring buffer.
+ */
+typedef struct {
+	TfwApmRBuf	rbuf;
+} TfwApmData;
+
+/*
+ * Note: This structure is constructed to fit into 32-bit type. If
+ * the basic constants of the algorithm change (see TFW_STATS_BCKTS,
+ * TFW_STATS_RANGES), then this structure will need a revision.
+ *
+ * @rtmval	Response time value.
+ * @intrvl	The interval number.
+ * @range	The range number.
+ * @bucket	The bucket number.
+ */
+typedef struct {
+	u16	rtmval;
+	u8	intrvl;
+	u8	range  : 4;
+	u8	bucket : 4;
+} __attribute__((packed)) TfwPcntCalc;
+
+/*
+ * [1ms, 349ms] should be sufficient for almost any installation,
+ * including cross atlantic.
+ */
+static const TfwPcntCtl __read_mostly pcntrng_ctl_init[TFW_STATS_RANGES] = {
+	{{0, 1, 16}},
+	{{1, 17, 47}},
+	{{2, 48, 108}},
+	{{4, 109, 349}}
+};
+
+static int tfw_apm_jtmwindow;		/* Time window in jiffies. */
+static int tfw_apm_jtmintrvl;		/* Time interval in jiffies. */
+static int tfw_apm_tmwscale;		/* Time window scale. */
+
+
+static void
+tfw_stats_calc(TfwPcntCalc *pcalc, size_t pcsz,
+	       TfwPcntRanges **rnglst, size_t rlsz,
+	       Percentile *pcntl, size_t plsz)
+{
+	int i, p = 0;
+	unsigned long cnt = 0, total_cnt = 0;
+	unsigned long pval[plsz];
+
+	/* Total hits count in all interval entries. */
+	for (i = 0; i < rlsz; ++i)
+		total_cnt += atomic64_read(&rnglst[i]->total_cnt);
+	if (unlikely(!total_cnt))
+		return;
+
+	/* The number of items to collect for each percentile. */
+	for (i = 0; i < plsz; ++i) {
+		pval[i] = total_cnt * pcntl[i].ith / 100;
+		if (!pval[i])
+			pcntl[p++].val = 0;
+	}
+	for (i = 0; i < pcsz; ++i) {
+		TfwPcntCalc *pc = pcalc + i;
+		cnt += atomic_read(&rnglst[pc->intrvl]->cnt[pc->range][pc->bucket]);
+		for ( ; p < plsz && pval[p] <= cnt; ++p)
+			pcntl[p].val = pc->rtmval;
```

Comment:
The only one thing where the algorithm differs from the unit test is ring buffer and essentially different ranges in its items. So basically we can just add one extra loop over the ring buffer (note that I proposed to have zero ranges) to `tfw_stat_calc()` from the unit test. Lets introduce an array `ptrs` of size equal to size of the ring buffer. The array must contain items like

         struct {
                   int r; // current i'th rb entry range
                   int b; // current i'th rb entry bucket
         }

The array is initialized by zeros. On first iteration we search minimal response time for 0th bucket or 0th range over all ring buffer (rb) entries and use the bucket for current `pcnts` calculation (from original unit test `tfw_stat_calc()`). `b` and probably `r` for current rb entry is moved forward. We repeat the search of the minimal bucket at second iteration, but we consider the next bucket of just used rb entry. The process repeats using only minimal buckets. That's long process, but it shorter, than current merging. We can thik about the algorithm as generalization of merge sort.